pkg/process/monitor_linux.go

```diff
@@ -0,0 +1,259 @@
+//go:build linux
+// +build linux
+
+/*
+   Copyright The nydus Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package process
+
+import (
+	"net"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/nydus-snapshotter/pkg/errdefs"
+	"github.com/containerd/nydus-snapshotter/pkg/utils/retry"
+)
+
+// LivenessMonitor liveness of a nydusd daemon.
+type LivenessMonitor interface {
+	// Subscribe death event of a nydusd daemon.
+	// `path` is where the monitor is listening on.
+	Subscribe(id string, path string, notifier chan<- deathEvent) error
+	// Unsubscribe death event of a nydusd daemon.
+	Unsubscribe(id string) error
+	// Run the monitor, wait for nydusd death event.
+	Run()
+	// Stop the monitor and release all the resources.
+	Destroy()
+}
+
+type target struct {
+	// A connection to the nydusd. Should close it when stopping the liveness monitor!
+	uc *net.UnixConn
+	// Notify subscriber that the nydusd is dead via the channel
+	notifier chan<- deathEvent
+	// `id` is usually the daemon ID
+	id   string
+	path string
+}
+
+type FD = uintptr
+
+type livenessMonitor struct {
+	mu sync.Mutex
+	// Get a subscribing target by the target ID (usually is the daemon ID)
+	subscribers map[string]*target
+	// Get a subscribing target by the connection FD. Each liveness
+	// probe has a unique connection FD which is listened for epoll event.
+	set     map[FD]*target
+	epollFd int
+}
+
+type deathEvent struct {
+	daemonID string
+	path     string
+}
+
+func newMonitor() (_ *livenessMonitor, err error) {
+	var epollFd int
+	epollFd, err = unix.EpollCreate1(unix.EPOLL_CLOEXEC)
+	if err != nil {
+		return nil, errors.Wrap(err, "create daemons monitor")
+	}
+
+	defer func() {
```

Comment:
Is it redundant?

Reply:
Do you mean it's not possible that `newMonitor` (this method) can't fail after L82?
Yes, it can't. Let's remove the error handler

Reply:
fixed